A83-xds-gcp-authn-filter.md

```diff
@@ -0,0 +1,218 @@
+the parsed cluster resource that is passed to the XdsClient watcher.
+
+The metadata field is a message that actually contains two maps:
+- [`filter_metadata`](https://github.com/envoyproxy/envoy/blob/7436690884f70b5550b6953988d05818bae3d087/api/envoy/config/core/v3/base.proto#L248):
+  This map contains `google.protobuf.Struct` values, which we will
+  convert to parsed JSON form.
+- [`typed_filter_metadata`](https://github.com/envoyproxy/envoy/blob/7436690884f70b5550b6953988d05818bae3d087/api/envoy/config/core/v3/base.proto#L257):
+  This map contains `google.protobuf.Any` fields.  To support this, we
+  will use a registry-like approach for metadata types (may be an actual
+  registry or just a block of code that supports the known protobuf
+  message types) that handles parsing the `google.protobuf.Any` field
+  and converting it some internal form appropriate for the implementation
+  (e.g., a native struct).
+
+The value for a given metadata key will come from only one of the
+two maps; the value from `filter_config` will be used only if the
+key is not present or is of an unknown protobuf message type in
+`typed_filter_config`.  In the resulting map in the parsed cluster
+resource, the map value will be either JSON or the appropriate internal
+form, depending on which of the two maps the entry came from.
+
+To match Envoy, the logic to validate cluster metadata will look something
+like this (pseudo-code):
+
+```
+parsed_metadata = {}  # Value is either JSON or parsed object
+all_keys = (cluster_metadata.typed_filter_metadata.keys() +
+            cluster_metadata.filter_metadata.keys())
+for key in all_keys:
+  # First try typed_filter_metadata.
+  any_field = cluster_metadata.typed_filter_metadata[key]
+  if any_field is not None:
+    parser = metadata_registry.FindParser(any_field.type_url)
+    if parser is not None:
+      value = parser.Parse(any_field.value)
+      if value is None:
+        return NACK  # Parsing failed, reject resource
+      parsed_metadata[key] = value
+      continue
+    # If no parser registered for this type, ignore the entry.
+  # Didn't find it in typed_filter_metadata, so try filter_metadata.
+  struct_field = cluster_metadata.filter_metadata[key]
+  if struct_field is not None:
+    parsed_metadata[key] = ConvertToJson(struct_field)
+```
+
+For now, the only registered metadata type we support is
+[`extensions.filters.http.gcp_authn.v3.Audience`](https://github.com/envoyproxy/envoy/blob/c16faca3619fb44c24b12d15aad8a797b9e210ab/api/envoy/extensions/filters/http/gcp_authn/v3/gcp_authn.proto#L66).
+In this message, the `url` field must be non-empty; if empty, the
+resource will be NACKed.  The parsed representation of this message can
+be a simple string.
+
+Note that in Envoy, the metadata keys must exactly match the legacy
+filter name (e.g., "envoy.filters.http.gcp_authn").  However, as per
+envoyproxy/envoy#34251, it is desirable to instead use the HTTP filter
```

Comment:
We already support Metadata keys that aren't tied to any filter. So this is only related to filters with specific cluster configuration?